json-rpc/src/tests/utils.rs

```diff
@@ -148,9 +148,10 @@ impl DbReader for MockDiemDB {
         &self,
         start_version: u64,
         limit: u64,
-        _ledger_version: u64,
+        ledger_version: u64,
         fetch_events: bool,
     ) -> Result<TransactionListWithProof, Error> {
+        let limit = std::cmp::min(limit, ledger_version - start_version + 1);
```

Comment:
Shouldn't we raise error if start_version + limit > ledger_version?

Reply:
Why would we return an error? If I ask for the next 1000 txs, but there are only 500 new txs, should we throw an error or just return the 500 new txs? The `get_transactions_with_proofs` doesn't let the user specify the `ledger_version`, so the user doesn't know  the current ledger version (which would define the max limit).

Reply:
I'd expect JSON-RPC API handler will do input param validation, thus all parameters passed into DB layer should be safe.
Thus in this test utility code, we should assert the parameters are safe, not fix them.

Reply:
Hmm, I'd actually go the other way, json-rpc is "just" another client of diemdb, so I wouldn't expect all db clients to need to repeat the same input validation, since db already correctly handles input validation in this case. For example, we can remove the `if start_version > ledger_version || limit == 0 { return Ok(TransactionListView::empty()) }`

Reply:
For example, you can see here the `get_transactions` handler becomes super simple:

```rust
/// Returns transactions by range
pub fn get_transactions(
    db: &dyn DbReader,
    ledger_version: u64,
    start_version: u64,
    limit: u64,
    include_events: bool,
) -> Result<TransactionListView, JsonRpcError> {
    let txs = db.get_transactions(start_version, limit, ledger_version, include_events)?;
    Ok(TransactionListView::try_from(txs)?)
}
```

Reply:
1. The error raised from DB is not well defined. If you want db handles all of these, you need make sure db interface does what you described before you make the change. 
2. You can keep adding layer for sharing logic, but somewhere need handle these logic. Current method does all input validation and response proper jsonrpc error. 
3. I don't really care where we put these logic, but it should be consistent.

Reply:
actually, there is layer created for this purpose already, it is data.rs in same directory with method.rs.

Reply:
I see what you're saying. The DB layer probably needs some work to be more consistent with its input validation and error handling. The interface stability is also a good point. In lieu of improving the db interface, I've added the checking back and some assertions in the mock db to make sure inputs are validated.